src/main/java/com/example/newslettercore/domain/user/service/UserService.java

```diff
@@ -1,42 +1,47 @@
 package com.example.newslettercore.domain.user.service;
 
-import com.example.newslettercore.domain.exception.CantLoginException;
 import com.example.newslettercore.domain.exception.NewsletterCoreObjectNotFoundException;
+import com.example.newslettercore.domain.port.PasswordEncipher;
+import com.example.newslettercore.domain.user.model.Role;
 import com.example.newslettercore.domain.user.model.User;
 import com.example.newslettercore.domain.user.repository.UserRepository;
 import com.example.newslettercore.domain.user.value.UserEmailValue;
 import com.example.newslettercore.domain.user.value.UserNameValue;
 import com.example.newslettercore.domain.user.value.UserPasswordValue;
+import lombok.RequiredArgsConstructor;
 import org.springframework.stereotype.Service;
 
+import java.util.Optional;
+
 @Service
+@RequiredArgsConstructor
 public class UserService {
 
     private final UserRepository userRepository;
+    private final PasswordEncipher passwordEncipher;
 
-    public UserService(UserRepository userRepository) {
+    public User createUser(UserNameValue name, UserPasswordValue password, UserEmailValue email, Role role) {
 
-        this.userRepository = userRepository;
+        User userToCreate = new User(name.getValue(), password.getValue(), email.getValue(), role);
+        return saveUser(userToCreate);
     }
 
-    public User createUser(UserNameValue name, UserPasswordValue password, UserEmailValue email) {
+    private User saveUser(User user) {
 
-        User userToCreate = new User(name.getValue(), password.getValue(), email.getValue());
-        return userRepository.save(userToCreate);
+        user.hashPassword(passwordEncipher::encodePassword);
+        return userRepository.save(user);
     }
 
     public User updateUser(String userId, UserNameValue name, UserPasswordValue password, UserEmailValue email) {
 
         User foundUser = userRepository.findById(userId)
                 .orElseThrow(() -> new NewsletterCoreObjectNotFoundException(User.class.getSimpleName(), userId));
         User updatedUser = foundUser.updateUser(name.getValue(), password.getValue(), email.getValue());
-        return userRepository.save(updatedUser);
+        return saveUser(updatedUser);
```

Comment:
Czy jak password nie zostanie zmieniony, to czy nie zostanie dwa razy zahashowane ? Tj. hasło wyciągnięte z bazy jest zahasowane i potem je hashujemy jeszcze raz. Jeżeli tak, to ja proponuje wydzielić zmiane hasła jak oddzielne flow od zmiany pozostałych danych, tak jak jest to większości systemów, jest to raczej specjalny przypadek. Da się wtedy ominąć jakieś sprawdzenia czy hasło zostało zmienione.

Reply:
Tak jak piszesz zostanie dwa razy zahashowane.